### PR TITLE
tracker allow list

### PR DESCRIPTION
**Pull request by jdorweiler**
**Reviewer:**

## Description:
Adding the ability to allowlist single trackers to temporarily fix site breakage.

## Steps to test this PR:
1. need the submodule tests here: https://github.com/duckduckgo/privacy-reference-tests/pull/5
2. `npm run test`

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/allowlisted-trackers.es6.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+const tdsStorage = require('./storage/tds.es6')
+const utils = require('./utils.es6')
+
+function isTrackerAllowlisted (site, request) {
+    if (tdsStorage.config.features.trackerAllowlist.state === 'disabled') {
+        return false
+    }
+
+    if (!tdsStorage.config.features.trackerAllowlist &&
+        !tdsStorage.config.reatures.trackerAllowlist.settings &&
```
Comment (sammacbeth) [inline shared/js/background/allowlisted-trackers.es6.js:10]:
```suggestion
        !tdsStorage.config.features.trackerAllowlist.settings &&
```

File: shared/js/background/allowlisted-trackers.es6.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if (allowListEntry) {
+        return _matchesRule(site, request, allowListEntry)
+    } else {
+        return false
+    }
+}
+
+function _findAllowlistEntry (request, allowList) {
+    const host = utils.extractHostFromURL(request).split('.')
+    const urlList = Array.from(host)
```
Comment (sammacbeth) [inline shared/js/background/allowlisted-trackers.es6.js:26]: Why do we need to copy this array?

File: shared/js/background/allowlisted-trackers.es6.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                matchedRule = ruleObj
+                return true
+            } else {
+                return false
+            }
+        })
+    }
+
+    if (matchedRule) {
+        if (matchedRule.domains.includes('<all>') || matchedRule.domains.includes(utils.extractHostFromURL(utils.extractLimitedDomainFromURL(site)))) {
```
Comment (sammacbeth) [inline shared/js/background/allowlisted-trackers.es6.js:55]: This checks for eTLD+1 in `matchedRule.domains` right? Why not just use `tldts.parse(site).domain`?

File: shared/js/background/redirect.es6.js [hunk 4]
```diff
@@ -120,6 +121,14 @@ function handleRequest (requestData) {
         }
 
         if (tracker) {
+            // temp allowlisted trackers to fix site breakage
+            const allowListed = trackerAllowlist(thisTab.site.url, requestData.url)
```
Comment (sammacbeth) [inline shared/js/background/redirect.es6.js:125]: I think there's also `thisTab.site.domain` which is already eTLD+1 which would save you needing to parse inside the `trackerAllowlist` function.

**Review by sammacbeth — CHANGES_REQUESTED** [review]

File: shared/js/background/allowlisted-trackers.es6.js (hunk 1 above)
Comment (jdorweiler) [reply shared/js/background/allowlisted-trackers.es6.js:10]: 🤦 thanks

File: shared/js/background/allowlisted-trackers.es6.js (hunk 3 above)
Comment (jdorweiler) [reply shared/js/background/allowlisted-trackers.es6.js:55]: I was trying to use the existing url parsing code. I don't mind pulling in tldts though.

File: shared/js/background/redirect.es6.js (hunk 4 above)
Comment (jdorweiler) [reply shared/js/background/redirect.es6.js:125]: `site.domain` can include the subdomain. I wanted allowlist matching to be on the root domain only. This way we don't need to worry about allowlisting every subdomain of a site that might be breaking.

File: shared/js/background/allowlisted-trackers.es6.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    const parsedRequest = new URL(request)
+    if (parsedRequest.port) {
+        request = request.replace(`:${parsedRequest.port}`, '')
+    }
+
+    if (allowListEntry.rules && allowListEntry.rules.length) {
+        allowListEntry.rules.some(ruleObj => {
+            if (request.match(ruleObj.rule)) {
+                matchedRule = ruleObj
+                return true
```
Comment (sammacbeth) [inline shared/js/background/allowlisted-trackers.es6.js:40]: IMO using `some` without using the return value is a bit of a confusing pattern - a for loop with a break would be easier to follow.

File: shared/js/background/redirect.es6.js [hunk 6]
```diff
@@ -120,6 +121,14 @@ function handleRequest (requestData) {
...
+            if (allowListed && utils.isFeatureEnabled('trackerAllowlist')) {
```
Comment (sammacbeth) [inline shared/js/background/redirect.es6.js:126]: 1. @jonathanKingston wrote a new helper to get this info from the `site` object.
 2. Can we check if the feature is enabled for the site before doing matching?

**Review by sammacbeth — COMMENTED** [review]
Looks good! I just have a couple of minor comments.

File: shared/js/background/allowlisted-trackers.es6.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+function _matchesRule (site, request, allowListEntry) {
+    let matchedRule = null
+    request = request.split('?')[0].split(';')[0]
+
+    // remove port from request urls
+    const parsedRequest = new URL(request)
+    if (parsedRequest.port) {
+        request = request.replace(`:${parsedRequest.port}`, '')
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/background/allowlisted-trackers.es6.js:33]: 👍 nice, good idea

**Review by sammacbeth — APPROVED** [review]
Looks good - just your unit tests are failing. Can you update the submodule commit so it pulls in the new tests?

**Pull request merged**